src/uharfbuzz/_harfbuzz.pyx

```diff
@@ -1057,3 +1057,152 @@ cdef class DrawFuncs:
         self._close_path_func = func
         hb_draw_funcs_set_close_path_func(
             self._hb_drawfuncs, _close_path_func, <void*>user_data, NULL)
+
+cdef class HBLinks:
+    cdef hb_link_t* _hb_link
+
+    cdef set(self, hb_link_t* p_link):
+        self._hb_link = p_link
+
+    def add_links_value(self, links: List[Tuple[int, int, int]]):
+        cdef unsigned int num_links = len(links)
+        for i in range(num_links):
+            self._hb_link[i].position = links[i][0]
+            self._hb_link[i].width = links[i][1]
+            self._hb_link[i].objidx = links[i][2]
+
+cdef hb_link_t* create_links(unsigned int link_num,
+                             hb_object_t* obj_list,
+                             unsigned int length,
+                             hb_bool_t* success):
+    success[0] = 1
+    cdef hb_link_t* p = NULL
+    if link_num == 0:
+        return NULL
```

Comment:
Is it intentional that success is not set to zero here?

Reply:
yes, when NULL is returned, there're 2 cased:
num of links = 0, success =1
num of links > 0 , success =0, means memory allocation failed.
So we need another boolean flag here to indicate whether memory allocating succeeds.